### ISSUE TITLE
How to get avarage result when apply k fold cross validation

### ISSUE THREAD

**Issue by lehieubkhn** [ultralytics/yolov5#3345]
## ❔Question
Hi, I'm currently training yolov5 model, applying 10 fold cross-validation. What I mean is train the model 10 times, 10 epochs each time, to choose the model with the best results and use all the data to train against that model to get the final result. But my instructor didn't agree with that direction, he told me to use the average result. So can you tell me how to calculate the average of 10 models like the way my teacher told me? Thanks

**Comment by glenn-jocher** [comment]
@Lehieubkhn see Model Ensembling tutorial, and you can modify the Ensemble method from `NMS` to `mean` here:
https://github.com/ultralytics/yolov5/blob/1f8d716ec943d9265cd33422d29560716e8b483c/models/experimental.py#L98-L111

## YOLOv5 Tutorials

* [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data)&nbsp; 🚀 RECOMMENDED
* [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)&nbsp; ☘️ RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Supervisely Ecosystem](https://docs.ultralytics.com/yolov5/tutorials/model_export8)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training)
* [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)&nbsp; ⭐ NEW
* [TorchScript, ONNX, CoreML Export](https://docs.ultralytics.com/yolov5/tutorials/model_export) 🚀
* [Test-Time Augmentation (TTA)](https://docs.ultralytics.com/yolov5/tutorials/test_time_augmentation)
* [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling)
* [Model Pruning/Sparsity](https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity)
* [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution)
* [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers)&nbsp; ⭐ NEW
* [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

**Comment by lehieubkhn** [comment]
> @Lehieubkhn see Model Ensembling tutorial, and you can modify the Ensemble method from `NMS` to `mean` here:
> https://github.com/ultralytics/yolov5/blob/1f8d716ec943d9265cd33422d29560716e8b483c/models/experimental.py#L98-L111
> 
> ## YOLOv5 Tutorials
> * [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data)  rocket RECOMMENDED
> * [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)  shamrock RECOMMENDED
> * [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)  star2 NEW
> * [Supervisely Ecosystem](https://docs.ultralytics.com/yolov5/tutorials/model_export8)  star2 NEW
> * [Multi-GPU Training](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training)
> * [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)  star NEW
> * [TorchScript, ONNX, CoreML Export](https://docs.ultralytics.com/yolov5/tutorials/model_export) rocket
> * [Test-Time Augmentation (TTA)](https://docs.ultralytics.com/yolov5/tutorials/test_time_augmentation)
> * [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling)
> * [Model Pruning/Sparsity](https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity)
> * [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution)
> * [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers)  star NEW
> * [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

Thanks in advance :D

**Comment by lehieubkhn** [comment]
@glenn-jocher  How the ensemble method work? With NMS, it's mean to take the box with the highest confidence score from 10 models? And with Mean is to take the average of all bounding boxes from 10 models? Am I right?

**Comment by glenn-jocher** [comment]
@Lehieubkhn yes that's correct.

**Comment by nandini211995** [comment]
> 
> ## ❔Question
> 
> Hi, I'm currently training yolov5 model, applying 10 fold cross-validation. What I mean is train the model 10 times, 10 epochs each time, to choose the model with the best results and use all the data to train against that model to get the final result. But my instructor didn't agree with that direction, he told me to use the average result. So can you tell me how to calculate the average of 10 models like the way my teacher told me? Thanks
> ## Additional context

Hi @Lehieubkhn , how did you perform 10 fold cross validation ? Did you split the dataset manually or apply any method?

**Comment by Zephyr69** [comment]
You were being asked for the cv score. The correct way you should do this is to get the outputs of all 10 folds separately, combine them as a single output, and then calculate the metrics. It would be nice to have this feature in the future.

There is no automatic way to do this in the repo yet; ensembling won't get you there as it generates a score for a single combined model on a single validation set.

**Comment by glenn-jocher** [comment]
@Lehieubkhn what are the typical results you see after training the 10 models for 10 epochs? I'm assuming you mean 10 models from scratch and not 10 pretrained models?

**Comment by qingyou97** [comment]
> ## ❔Question
> Hi, I'm currently training yolov5 model, applying 10 fold cross-validation. What I mean is train the model 10 times, 10 epochs each time, to choose the model with the best results and use all the data to train against that model to get the final result. But my instructor didn't agree with that direction, he told me to use the average result. So can you tell me how to calculate the average of 10 models like the way my teacher told me? Thanks
> 
> ## Additional context

hello,Can I ask how to use k-folds to  train? I want to learn!

**Comment by glenn-jocher** [comment]
@qingyou97 hi! To use k-fold cross-validation for training YOLOv5, you can split your dataset into k subsets and then train and validate the model k times, each time using a different subset as the validation set and the remaining subsets as the training set. Once you have the results from all k folds, you can calculate the average of the metrics to assess the model's performance. Keep in mind that the current YOLOv5 implementation does not have automatic support for k-fold cross-validation, but you can achieve this manually by organizing your data accordingly.
